### PR TITLE
8 feature 8 save player stats feature

### PR DESCRIPTION
**Pull request by lilyxmeng**
Created LoadPlayer and SavePlayer classes, along with SaveLoadPlayerTest test unit to check various cases. This PR adds data persistence to the project (save information and load that information later). The file to which information is being saved and loaded is player.csv, which is currently under the use_cases package. 

Concerns I would like addressed for review:
- Should player.csv be under use_cases or a lower layer? (Adhering to CA)
- Are there any edge cases that I've missed in my tests?

### CODE REVIEW

File: src/main/java/entities/HumanPlayer.java
Comment (nyan-ko) [inline src/main/java/entities/HumanPlayer.java:49]: this is technically unnecessary because of the getStats() method here, however i prefer distinct methods as opposed to putting everything into an array. we should decide which behaviour to stick with

File: src/main/java/use_cases/LoadPlayer.java
Comment (nyan-ko) [inline src/main/java/use_cases/LoadPlayer.java:1]: i feel like something as high level as a use case shouldnt be handling the low level details, which here would be reading a csv file. we should define an interface that supports loading and saving, then make an implementation of that interface that specifically handles csv reading and writing.

File: src/main/java/use_cases/LoadPlayer.java
Comment (nyan-ko) [inline src/main/java/use_cases/LoadPlayer.java:15]: just as an example:

```
public interface DataAccess {
    PlayerInformation load(String name, String file)  // here PlayerInformation would just be a data object of all the player stats
    void save(String name, String file)
}
```

with an example implementation:

```
public class CSVDatabase implements DataAccess {
  ...
  
  @Override
  public PlayerInformation load(String name, String file) {
    BufferedReader br = ...
    // your existing code in LoadPlayer
  }

  @Override
  public void save(String name, String file) {
    // your code in SavePlayer
  }
}
```
your existing usecases would then store a reference to a DataAccess interface and call its method, and we'd pass in a CSVDatabase class wherever we initialize them

the specifics can be decided later but i think this is a good guideline for keeping use cases separate from the databases.

File: src/main/java/use_cases/SavePlayer.java
Comment (nyan-ko) [inline src/main/java/use_cases/SavePlayer.java:1]: same comment for LoadPlayer

**Review by nyan-ko — CHANGES_REQUESTED** [review]

**Comment by nyan-ko** [comment]
player.csv should be at a lower level than use cases. following clean architecture guidelines, the csv file is a low level detail and should be in the outermost layer, however i think we should package it with the rest of the csv database files, for example, in a `database` folder

tests look comprehensive to me 👍

File: src/main/java/entities/HumanPlayer.java
Comment (SMateen21) [reply src/main/java/entities/HumanPlayer.java:49]: This behaviour is better I believe.

File: src/main/java/use_cases/SavePlayer.java
Comment (SMateen21) [inline src/main/java/use_cases/SavePlayer.java:17]: How/where would we handle this exception? That wouldn't be included here but just a thought.

File: src/main/java/use_cases/SavePlayer.java
Comment (SMateen21) [inline src/main/java/use_cases/SavePlayer.java:1]: I'm not too sure about including this in the use_cases packages, it is too concerned with the way we are saving Players (the "how"). I'd go with Edric's suggestion here.

**Review by SMateen21 — COMMENTED** [review]
Looks good! There is a Clean Architecture violation that can be avoided, I'd go with Edric's suggestion there before deciding to merge.

File: src/main/java/entities/HumanPlayer.java
Comment (lilyxmeng) [reply src/main/java/entities/HumanPlayer.java:49]: Okay I'll be removing getStats(). There are a few tests in HumanPlayerTest that use the getStats() method so I'll be fixing it in my following commits.

File: src/main/java/use_cases/SavePlayer.java
Comment (lilyxmeng) [reply src/main/java/use_cases/SavePlayer.java:1]: I had to include the use_cases package here because SavePlayer is in the use_cases package.

File: src/main/java/use_cases/LoadPlayer.java
Comment (lilyxmeng) [reply src/main/java/use_cases/LoadPlayer.java]: I've pushed new changes according to this suggestion. Let me know if I should make further edits.

File: src/main/java/use_cases/CSVDatabase.java
Comment (SMateen21) [inline src/main/java/use_cases/CSVDatabase.java:1]: I'd change this package to be something like "drivers" or "database" , something like that, but other than that all is good.

**Review by SMateen21 — APPROVED** [review]
Looks great! The DataAccess interface will make it easier for use case classes to load players and save them. Only thing would be changing the package which CSVDatabase is under though, as I don't think it should be part of use_cases.

File: src/main/java/use_cases/SavePlayer.java
Comment (lilyxmeng) [reply src/main/java/use_cases/SavePlayer.java]: The method will throw an IOException if it has an error reading or accessing the file. I don't think this is likely to happen given our relatively simple use of saving/loading data, but I could change this to a catch statement and explicitly handle it although I'm not sure what we would do.

File: src/main/java/use_cases/CSVDatabase.java
Comment (nyan-ko) [reply src/main/java/use_cases/CSVDatabase.java:1]: agreed

File: src/main/java/use_cases/DataAccess.java
Comment (nyan-ko) [inline src/main/java/use_cases/DataAccess.java:14]: i think a data class instead of a string array is better for this return value

File: src/main/java/use_cases/CSVDatabase.java
Comment (lilyxmeng) [reply src/main/java/use_cases/CSVDatabase.java:1]: That makes sense, I'll make that change.

File: src/main/java/use_cases/DataAccess.java
Comment (lilyxmeng) [reply src/main/java/use_cases/DataAccess.java]: Got it, created PlayerInformation class.

File: src/test/java/database/SaveLoadPlayerTest.java
Comment (nyan-ko) [inline src/test/java/database/SaveLoadPlayerTest.java:20]: add access modifier

Comment (lilyxmeng) [reply src/test/java/database/SaveLoadPlayerTest.java]: Fixed

**Review by nyan-ko — APPROVED** [review]
LGTM!

File: src/main/java/database/CSVDatabase.java
Comment (meiiph) [inline src/main/java/database/CSVDatabase.java:32]: save* (instead of saves)

File: src/main/java/database/DataAccess.java
Comment (meiiph) [inline src/main/java/database/DataAccess.java:18]: same comment as the one I just made, save* rather than saves :)

File: src/main/java/usecases/PlayerInformation.java
Comment (meiiph) [inline src/main/java/usecases/PlayerInformation.java:1]: Anywhere "this players" appears should be replaced with "this player's" in this file.

File: src/test/java/database/SaveLoadPlayerTest.java
Comment (meiiph) [inline src/test/java/database/SaveLoadPlayerTest.java:130]: Is this supposed to save or not save the player? The wording is a bit confusing here.

**Review by meiiph — APPROVED** [review]
Good, just fix the Javadocs comments that I added. :)

**Review by SMateen21 — APPROVED** [review]
Everything looks good. I'd add some light Javadocs for the classes as well before merging but otherwise I approve.

**Pull request merged**
